### PR TITLE
noImplicitAny: request.ts

### PR DESCRIPTION
**Pull request by JCIV-SE**
Fixes #<issue_number_goes_here> (it's a good idea to open an issue first for discussion)

- [ ] Tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: src/request.ts [hunk 1]
```diff
@@ -270,7 +191,7 @@ class DatastoreRequest {
             keys:
                 new Array(options.allocations).fill(entity.keyToKeyProto(key)),
           },
-          gaxOpts: options.gaxOptions,
+          gaxOpts: undefined,
```
Comment (praveenqlogic) [inline src/request.ts]: can we use empty object here ie. {}

File: src/request.ts [hunk 2]
```diff
@@ -306,25 +227,14 @@ class DatastoreRequest {
...
-    if (keys.length === 0) {
-      throw new Error('At least one Key object is required.');
-    }
-
-    const makeRequest = keys => {
-      // tslint:disable-next-line no-any
-      const reqOpts: any = {
+  createReadStream(key: Keys): Stream;
+  createReadStream(key: Keys, options: CreateReadStreamOptions): Stream;
+  createReadStream(keys: Keys, options: CreateReadStreamOptions = {}): Stream {
```
Comment (praveenqlogic) [inline src/request.ts]: options will be optional

File: src/request.ts [hunk 3]
```diff
@@ -424,20 +334,13 @@ class DatastoreRequest {
...
-      callback = gaxOptions;
-      gaxOptions = {};
-    }
-
-    callback = callback || (() => {});
+  delete(keys: Keys): void;
+  delete(keys: Keys, gaxOptions?: DeleteGaxOptions): void|Promise<DeleteResponse>;
+  delete(keys: Keys, callback?: DeleteCallback): void;
+  delete(keys: Keys, gaxOptions?: DeleteGaxOptions, callback?: DeleteCallback): void;
+  delete(keys: Keys, gaxOptionsOrCallback?: DeleteGaxOptions|DeleteCallback, cb?: DeleteCallback): void| Promise<DeleteResponse> {
```
Comment (praveenqlogic) [inline src/request.ts]: delete(keys: Keys): void; - not required
delete(keys: Keys, gaxOptions?: DeleteGaxOptions):Promise<DeleteResponse>; no void!
delete(keys: Keys, callback: DeleteCallback): void; - callback should not optional
delete(keys: Keys, gaxOptions: DeleteGaxOptions, callback: DeleteCallback): void - no optional

File: src/request.ts [hunk 4]
```diff
@@ -547,22 +450,27 @@ class DatastoreRequest {
...
-  get(keys: Entity|Entity[], options?: CreateReadStreamOptions|undefined): Promise<EntityDataObj[]>;
-  get(keys: Entity|Entity[], options?: CreateReadStreamOptions|undefined, callback?:(a:null, b:boolean|number|string|Array<number|string|boolean>) => void): void|Promise<EntityDataObj[]> {
-    if (is.fn(options)) {
-      callback = options;
-      options = {};
+  //! Refactor for Proto File
+  get(keys: Keys, options?: GetOptions): Entity | Stream;
+  get(keys: Keys, callback?: GetCallback): Entity | Stream;
+  get(keys: Keys, options?: GetOptions, callback?: GetCallback): Entity | Stream;
+  get(keys: Keys, optionsOrCallback?: GetOptions|GetCallback, cb?: GetCallback): Entity | Stream {
```
Comment (praveenqlogic) [inline src/request.ts]: get(keys: Keys, options?: GetOptions): Promise<Entity | Stream>;
get(keys: Keys, callback: GetCallback): void;
get(keys: Keys, options: GetOptions, callback: GetCallback): void;
get(keys: Keys, optionsOrCallback?: GetOptions|GetCallback, cb?: GetCallback):void | promise<Entity|Stream>;

please verify once EntityDataObj[] do, you may use this to return from function!

File: src/request.ts [hunk 5]
```diff
@@ -582,36 +490,18 @@ class DatastoreRequest {
...
-  //! InsertCallback
-  insert(entities: Entity|Entity[], callback: CallOptions): void {
-    entities =
-        arrify(entities).map(DatastoreRequest.prepareEntityObject_).map(x => {
-          x.method = 'insert';
-          return x;
-        });
-
+  insert(entities: Entities): Promise<InsertResponse>;
+  insert(entities: Entities, callback?: InsertCallback): void;
```
Comment (praveenqlogic) [inline src/request.ts]: callback should optional

File: src/request.ts [hunk 6]
```diff
@@ -761,19 +651,11 @@ class DatastoreRequest {
...
-      //! ConsistencyProtoCode has implicit any type
-  //* Error object
-  //! Response Object
-  //! RunQueryStreamResponse (promise)
-  //! RunQueryStreamOptions
-  //! makeRequest "query" annotation
-  runQueryStream(query: Query, options?) {
-    options = options || {};
+  runQueryStream(query: Query, options?: RunQueryStreamOptions): Stream;
+  runQueryStream(query: Query, options: RunQueryStreamOptions = {}): Stream {
```
Comment (praveenqlogic) [inline src/request.ts]: hmmm, if options got a default value, do we need a overloading for optional options?, please correct me also!

File: src/request.ts [hunk 7]
```diff
@@ -1066,30 +948,29 @@ class DatastoreRequest {
...
-  //! this.request_ "Object is possibly undefined"
-  //! this.requestCallbacks_ "Does not exist on this"
-  //! "x" annotation
-  save(entities: Entity[], gaxOptions?: CallOptions, callback?):void {
+  //! Overloading needs attention
+  save(entities: Entities): void;
+  save(entities: Entities, callback?: SaveCallback): void;
+  save(entities: Entities, gaxOptions?: SaveOptions): void|Promise<SaveResponse>;
+  save(entities: Entities, gaxOptionsOrCallback?: SaveOptions, callback?: SaveCallback): void;
+  save(entities: Entities, gaxOptionsOrCallback?: SaveOptions|SaveCallback, cb?: SaveCallback): void|Promise<SaveResponse> {
```
Comment (praveenqlogic) [inline src/request.ts]: save(entities: Entities, gaxOptions?: SaveOptions): Promise<SaveResponse>;
save(entities: Entities, callback: SaveCallback): void;
save(entities: Entities, gaxOptions: SaveOptions, callback: SaveCallback): void;
save(entities: Entities, gaxOptionsOrCallback?: SaveOptions|SaveCallback, cb?: SaveCallback): void|Promise<SaveResponse>

i think its better if we use array of Entity[] in place of entities type

File: src/request.ts [hunk 8]
```diff
@@ -1121,15 +1001,15 @@ class DatastoreRequest {
           // This was replaced with a more efficient mechanism in the top-level
           // `excludeFromIndexes` option.
           if (is.array(entityObject.data)) {
-            entityProto.properties = entityObject.data.reduce((acc, data) => {
+            entityProto.properties = entityObject.data.reduce((acc: {[key: string]: ValueProto}, data: { value: ValueProto; excludeFromIndexes: ValueProto; name: string|number; }) => {
```
Comment (praveenqlogic) [inline src/request.ts]: look better if you encapsulate it and create a type/interface.

File: src/request.ts [hunk 9]
```diff
@@ -1206,15 +1086,11 @@ class DatastoreRequest {
...
-      //! UpdateResponse (promise)
-  //! Entities
-  //! UpdateCallback
-  //? TSlint not picking up on certain "any's", should I annotate anyway?
-  //? Overload
-  update(entities, callback) {
-    entities =
-        arrify(entities).map(DatastoreRequest.prepareEntityObject_).map(x => {
+  update(entities: Entities, callback: UpdateCallback): void|Promise<UpdateResponse>; //! Promise might be redundant
+  update(entities: Entities, callback: UpdateCallback): void|Promise<UpdateResponse> {
```
Comment (praveenqlogic) [inline src/request.ts]: update(entities: Entities): Promise<UpdateResponse>;
update(entities: Entities, callback: UpdateCallback): void;
update(entities: Entities, callback?: UpdateCallback): void|Promise<UpdateResponse>;

File: src/request.ts [hunk 10]
```diff
@@ -1236,14 +1112,11 @@ class DatastoreRequest {
...
-  //* Void return annotation
-      //! UpsertResponse (promise)
-  //! Entities
-  //! UpsertCallback
-  //? Overload
-  upsert(entities: Entity[], callback): void {
-    entities =
-        arrify(entities).map(DatastoreRequest.prepareEntityObject_).map(x => {
+  upsert(entities: Entities, callback: UpsertCallback): void|Promise<UpsertResponse>; //! Promise might be redundant
+  upsert(entities: Entities, callback: UpsertCallback): void|Promise<UpsertResponse> {
```
Comment (praveenqlogic) [inline src/request.ts]: upsert(entities: Entities)Promise&lt;UpsertResponse&gt;
upsert(entities: Entities, callback: UpsertCallback): void;
upsert(entities: Entities, callback?: UpsertCallback): void|Promise&lt;UpsertResponse&gt;

File: src/request.ts [hunk 11]
```diff
@@ -1263,12 +1136,8 @@ class DatastoreRequest {
...
    * @private
    */
-  //* Needs return annotation
-      //! Needs Promise return annotation
-  //! RequestConfig
-  //! RequestCallback
-  //? Overloading
-  request_(config: RequestConfig, callback: RequestCallback): void {
+  request_(config: RequestConfig, callback: RequestCallback): void|Promise<RequestResponse>;
+  request_(config: RequestConfig, callback: RequestCallback): void|Promise<RequestResponse> {
```
Comment (praveenqlogic) [inline src/request.ts]: request_(config: RequestConfig): void;
request_(config: RequestConfig, callback: RequestCallback): void;
request_(config: RequestConfig, callback?: RequestCallback): void|Promise&lt;RequestResponse&gt;

File: system-test/datastore.ts [hunk 12]
```diff
@@ -48,8 +48,10 @@ describe('Datastore', () => {
   it('should allocate IDs', done => {
     datastore.allocateIds(datastore.key('Kind'), 10, (err, keys) => {
       assert.ifError(err);
-      assert.strictEqual(keys.length, 10);
-      assert.strictEqual(entity.isKeyComplete(keys[0]), true);
+      if (keys !== null) {
```
Comment (praveenqlogic) [inline system-test/datastore.ts]: hey, you need to workaround this, the idea of this test case looks like to check keys exists

File: src/entity.ts [hunk 13]
```diff
@@ -18,6 +18,7 @@ import * as arrify from 'arrify';
 import * as extend from 'extend';
 import * as is from 'is';
 import {Query, QueryProto} from './query';
+// import {google} from '../proto/datastore';
```
Comment (nareshqlogic) [inline src/entity.ts]: Remove comment line if no needed

File: src/request.ts [hunk 14]
```diff
@@ -24,27 +24,29 @@ import * as is from 'is';
 import {split} from 'split-array-stream';
 import * as streamEvents from 'stream-events';
 import * as through from 'through2';
+import { google } from '../proto/datastore';
+// import { googleEntity } from '../proto/entity';
```
Comment (nareshqlogic) [inline src/request.ts]: Please remove comment line if no needed

File: src/request.ts [hunk 15]
```diff
@@ -24,27 +24,29 @@ import * as is from 'is';
...
 
-import {entity, Entity} from './entity';
+import {entity, Entity, ResponseResult, KeyProto, ValueProto, EntityProto} from './entity';
 import {Query, RunQueryInfo, RunQueryOptions, RunQueryResponse, RunQueryCallback} from './query';
 import {Datastore} from '.';
-
-export interface EntityDataObj {
-  [key: string]: string;
-}
+import { Enum } from 'protobufjs';
```
Comment (nareshqlogic) [inline src/request.ts]: Please remove this import statement.

File: src/request.ts [hunk 16]
```diff
@@ -24,27 +24,29 @@ import * as is from 'is';
...
+import { isObject } from 'util';
```
Comment (nareshqlogic) [inline src/request.ts]: Please remove this import statement.

File: src/request.ts [hunk 17]
```diff
@@ -24,27 +24,29 @@ import * as is from 'is';
...
+import {CallOptions, GoogleError} from 'google-gax';
+import { Stream } from 'stream';
 
 // Import the clients for each version supported by this package.
 const gapic = Object.freeze({
   v1: require('./v1'),
 });
 
-import {entity, Entity} from './entity';
+import {entity, Entity, ResponseResult, KeyProto, ValueProto, EntityProto} from './entity';
```
Comment (nareshqlogic) [inline src/request.ts]: Please remove ResponseResult and EntityProto.

File: src/request.ts [hunk 18]
```diff
@@ -24,27 +24,29 @@ import * as is from 'is';
 import {split} from 'split-array-stream';
 import * as streamEvents from 'stream-events';
 import * as through from 'through2';
+import { google } from '../proto/datastore';
+// import { googleEntity } from '../proto/entity';
+import {CallOptions, GoogleError} from 'google-gax';
```
Comment (nareshqlogic) [inline src/request.ts]: Please remove GoogleError from import statement as no used anywhere.

File: src/request.ts (hunk 1 above)
Comment (JCIV-SE) [reply src/request.ts]: I tried that, actually, good suggestion. I fail a test either way so I changed it back to options.gaxOptions. Was originally changed to pass TSLint before I optimized for the proto file

File: system-test/datastore.ts (hunk 12 above)
Comment (praveenqlogic) [reply system-test/datastore.ts]: keys is possibly 'null' for this try **keys!**

**Pull request merged**